tree-sitter/src/TreeSitter/Unmarshal.hs

```diff
@@ -260,48 +298,67 @@ instance GUnmarshal U1 where
-
-class GUnmarshalSum f where
-  gunmarshalSumNode :: (MonadFail m
-                   , Carrier sig m
-                   , Member (Reader ByteString) sig
-                   , Member (Reader (Ptr Cursor)) sig
-                   , MonadIO m)
-                   => Node -> m (f a)
+  gunmarshalNode node = push getFields >>= gunmarshalProductNode @(f :*: g) node . fromMaybe Map.empty
 
-instance (Unmarshal k, SymbolMatching k) => GUnmarshalSum (M1 C c (M1 S s (K1 i k))) where
-  gunmarshalSumNode node = M1 . M1 . K1 <$> unmarshalNodes [node]
-
-instance (GUnmarshalSum f, GUnmarshalSum g, SymbolMatching f, SymbolMatching g) => GUnmarshalSum (f :+: g) where
-  gunmarshalSumNode node = do
+-- For sum datatypes:
+instance (GUnmarshal f, GUnmarshal g, SymbolMatching f, SymbolMatching g) => GUnmarshal (f :+: g) where
+  gunmarshalNode node = do
     let lhsSymbolMatch = symbolMatch (Proxy @f) node
         rhsSymbolMatch = symbolMatch (Proxy @g) node
-    if lhsSymbolMatch
-      then L1 <$> gunmarshalSumNode @f node
-      else if rhsSymbolMatch
-        then R1 <$> gunmarshalSumNode @g node
-        else fail $ showFailure (Proxy @f) node `sep` showFailure (Proxy @g) node
+    if lhsSymbolMatch then
+      L1 <$> gunmarshalNode @f node
+    else if rhsSymbolMatch then
+      R1 <$> gunmarshalNode @g node
+    else
+      fail $ showFailure (Proxy @f) node `sep` showFailure (Proxy @g) node
 
 
 -- | Generically unmarshal products
 class GUnmarshalProduct f where
-  gunmarshalProductNode :: (MonadFail m, Carrier sig m, Member (Reader ByteString) sig, Member (Reader (Ptr Cursor)) sig, MonadIO m) => Map.Map FieldName [Node] -> m (f a)
+  gunmarshalProductNode
+    :: ( Carrier sig m
+       , Member (Reader ByteString) sig
+       , Member (Reader (Ptr Cursor)) sig
+       , MonadFail m
+       , MonadIO m
+       , UnmarshalAnn a
+       )
+    => Node
+    -> Fields
+    -> m (f a)
 
 -- Product structure
 instance (GUnmarshalProduct f, GUnmarshalProduct g) => GUnmarshalProduct (f :*: g) where
-  gunmarshalProductNode fields = (:*:) <$> gunmarshalProductNode @f fields <*> gunmarshalProductNode @g fields
+  gunmarshalProductNode node fields = (:*:)
+    <$> gunmarshalProductNode @f node fields
+    <*> gunmarshalProductNode @g node fields
 
 -- Contents of product types (ie., the leaves of the product tree)
-instance (Unmarshal k, Selector c) => GUnmarshalProduct (M1 S c (K1 i k)) where
-  gunmarshalProductNode fields =
-   M1 . K1 <$> unmarshalNodes (fromMaybe [] (Map.lookup (FieldName (selName @c undefined)) fields))
+instance UnmarshalAnn k => GUnmarshalProduct (M1 S c (K1 i k)) where
+  gunmarshalProductNode node _ = M1 . K1 <$> unmarshalAnn node
+
+instance GUnmarshalProduct (M1 S c Par1) where
+  gunmarshalProductNode node _ = M1 . Par1 <$> unmarshalAnn node
+
+instance (UnmarshalField f, Unmarshal g, Selector c) => GUnmarshalProduct (M1 S c (f :.: g)) where
```

Comment:
Optional/multiple fields are now exclusively handled via this instance. This means that we could technically eliminate `UnmarshalField` altogether, replacing it with multiple specializations of this instance. However, it would be a little more challenging to share code between them, and it would needlessly constrain users who might want to define their own AST datatypes instead of deriving them from the grammar.